Duckarray tests constructors

TomNicholas opened the pull request:
Builds on top of #4972 to add tests for `Variable/DataArray/Dataset` constructors and properties when wrapping duck arrays.

- [x] Closes part of #6894
- [ ] Tests added (tests for tests?? Maybe...)
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

TomNicholas commented on xarray/tests/duckarrays/base/strategies.py:
```diff
@@ -0,0 +1,199 @@
+T_DuckArray = Any
+
+
+@st.composite
+def duckarray(
+    draw,
+    create_data,
+    *,
+    sizes=None,
+    min_size=1,
+    max_size=3,
+    min_dims=1,
+    max_dims=3,
+    dtypes=None,
+) -> st.SearchStrategy[T_DuckArray]:
```
I tried to add a new strategy for creating a duck array of varying shape/dtype, but ideally I would like to refactor the other strategies to depend on this one.

TomNicholas commented on xarray/tests/duckarrays/base/strategies.py:
```diff
@@ -0,0 +1,199 @@
+    return sizes
+
+
+# Is there a way to do this in general?
+# Could make a Protocol...
+T_DuckArray = Any
```
I added one or two type hints just as code documentation for myself, but I'm not sure if typing these base classes in general would yield any benefit in terms of catching errors with duck array integrations?

TomNicholas replied:
Using a protocol (see @Illviljan 's [links](https://github.com/pydata/xarray/issues/6894#issuecomment-1209756650)) would be perfect in theory but in practice xarray doesn't actually require all the attributes defined in the array API, only a subset, so perhaps we should maintain our own duck array protocol so as not to be overly strict...?

Illviljan replied:
I think you should move `T_Duckarray `to `xarray.core.types` and then import it from there. It should be useful in other places as well.

Yeah xarray may not use everything available in the array api (yet). But as you've noticed we may get a lot of hypothesis and typing tests figured out for us.

TomNicholas commented on xarray/tests/duckarrays/base/constructors.py:
```diff
@@ -0,0 +1,59 @@
+            strategies.duckarray(lambda shape, dtypes: self.create(shape, dtypes))
+        )
+
+        # TODO generalize to N dimensions
+        dims = ["dim_0", "dim_1", "dim_2"]
```
This is related to https://github.com/pydata/xarray/pull/4972#discussion_r941690041 because if dimension names aren't part of the strategies then we could just have a utility function which generates lists of dimension names of the form `dim_*`.

TomNicholas replied:
Oh the exact function I just wrote already exists as `utils.create_dimension_names`, my bad! :sweat_smile:

TomNicholas commented:
@keewis I've come back to this PR now that #8404 is merged. I have:
- Rewritten the reduce tests to use the new `variables` strategy
- Used the `array_strategy_fn` pattern everywhere
- Added a new file for testing xarray compatibility with the minimal implementation of the array API standard `numpy.array_api`
- Rewritten the sparse tests to match
- Removed the pint tests / DataArray tests for now

I just wanted to show you to see what you thought of this approach now.

TomNicholas commented on xarray/testing/duckarrays.py:
```diff
@@ -0,0 +1,140 @@
+    shapes = npst.array_shapes()
+    dtypes = xrst.supported_dtypes()
+
+    @staticmethod
+    @abstractmethod
+    def array_strategy_fn(
+        *, shape: "_ShapeLike", dtype: "_DTypeLikeNested"
+    ) -> st.SearchStrategy[T_DuckArray]:
+        # TODO can we just make this an attribute?
+        ...
+
```
I would like to just do this in the subclassed tests:

```python
class VariableConstructorTests

    array_strategy_fn = ...

```
but then would I need to make `array_strategy_fn` an abstract property?

TomNicholas commented on xarray/testing/duckarrays.py:
```diff
@@ -0,0 +1,140 @@
+
+        assert_identical(actual, expected)
+
+    @pytest.mark.parametrize(
+        "method",
+        (
+            "all",
+            "any",
+            # "cumprod",  # not in array API
+            # "cumsum",  # not in array API
+            # "max",  # only in array API for real numeric dtypes
+            # "max",  # only in array API for real floating point dtypes
+            # "median",  # not in array API
+            # "min",  # only in array API for real numeric dtypes
+            # "prod",  # only in array API for numeric dtypes
+            # "std",  # TypeError: std() got an unexpected keyword argument 'ddof'
+            # "sum",  # only in array API for numeric dtypes
+            # "var",  # TypeError: std() got an unexpected keyword argument 'ddof'
```
Clearly we need to be able to test certain reductions with only certain dtypes

TomNicholas commented on xarray/testing/duckarrays.py:
```diff
@@ -0,0 +1,140 @@
+    def test_construct(self, data) -> None:
+        shape = data.draw(self.shapes)
+        dtype = data.draw(self.dtypes)
+        arr = data.draw(self.array_strategy_fn(shape=shape, dtype=dtype))
```
You can't use `strategies.variables` to test the constructor because `strategies.variables` has a call to the constructor inside it, so you end up not being able to access the original raw array that you want to use as the "expected" value.

TomNicholas commented:
Okay I'm a bit stumped by how to do this. I want some test setup where:

1) We define sets of tests that a downstream library can import and run using just a couple of lines of code,
2) The downstream library can override the generated array type and the exact way equality is checked (e.g. to test sparse type arrays),
3) Certain tests are parametrized over many cases, which we ideally set upstream (e.g. over `sum`, `mean`, `std`, etc.),
4) But the downstream library can choose to xfail certain cases (e.g. because they haven't implemented `std` yet),
5) Different cases are tested for different dtypes automatically (e.g. because the array API standard supports string dtypes in `any` but not in `max`),
6) But the downstream library can also override those dtypes too (e.g. because they support more/less than what's currently in the array API).

I was hoping to do this using some combination of upstream test classes, `pytest.parametrize` on the test method, and `xfail`, but I can't really see how to get (4) and (6) to work with that approach. I might have to resort to using globals or something...

keewis commented:
If you look at #4972, there's a new mark that applies other marks, like `skip`. The API is a bit awkward, and might be out of date, but demonstrates that it is possible. Similarly, if you pass the dtypes as a fixture you should be able to override it on a per-module level (`pytest` has a whole bunch of docs on overriding fixtures).

TomNicholas commented:
> If you look at https://github.com/pydata/xarray/pull/4972, there's a new mark that applies other marks, like skip.

Interesting, thank you!

> Similarly, if you pass the dtypes as a fixture you should be able to override it on a per-module level (pytest has a whole bunch of docs on overriding fixtures).

Also a good idea, but I think I need to override the dtype fixtures on a per-parametrization level...

keewis commented:
Depends on what you want to parametrize, I guess. In general, though, I'm starting to lean towards not parametrizing the functions to test: `pytest.mark.parametrize` creates "variants" of the same test (i.e. it should test the same thing with different inputs), where having one function per "test" makes the summaries a bit more useful. Plus, running the tests of a single function are much easier that way compared to if they were variants. The downside is code duplication, but I don't think that's too much of an issue, especially for tests.
